templates/fine-tune-stable-diffusion/README.ipynb

```diff
@@ -0,0 +1,390 @@
+    "\n",
+    "A few notes:\n",
+    "* `ray.init(runtime_env={\"env_vars\": ...})` sets the environment variables on all workers in the cluster -- setting the environment variable in this notebook on the head node is not enough in a distributed setting.\n",
+    "* `train_fn_per_worker` is the function that will run on all distributed training workers. In this case, it's just a light wrapper on top of the `diffusers` example script that copies the latest checkpoint to shared cluster storage.\n",
+    "* `ScalingConfig` is the configuration that determines how many workers and what kind of accelerator to use for training. Once the training is launched, **Anyscale will automatically scale up nodes to meet this resource request!**\n",
+    "\n",
+    "The result of this fine-tuning will be a fine-tuned LoRA model checkpoint at `MODEL_CHECKPOINT_PATH`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "MODEL_CHECKPOINT_PATH = os.environ[\"ANYSCALE_ARTIFACT_STORAGE\"] + \"/checkpoint-final\"\n",
+    "\n",
+    "print(\"Final checkpoint will be uploaded to: \", MODEL_CHECKPOINT_PATH)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "import ray\n",
+    "import ray.train\n",
+    "from ray.train.torch import TorchTrainer\n",
+    "\n",
+    "from train_dreambooth_lora_sdxl import main\n",
+    "from utils import download_from_cloud, upload_to_cloud\n",
+    "\n",
+    "\n",
+    "# Set environment variables across the entire cluster.\n",
+    "ray.init(\n",
+    "    runtime_env={\n",
+    "        \"env_vars\": {\"WANDB_API_KEY\": os.environ.get(\"WANDB_API_KEY\")}\n",
+    "    },\n",
+    "    ignore_reinit_error=True,\n",
+    ")\n",
+    "\n",
+    "\n",
+    "def train_fn_per_worker(config: dict):\n",
+    "    download_from_cloud(cloud_uri=DATA_CLOUD_PATH, local_path=SUBJECT_IMAGES_PATH)\n",
+    "\n",
+    "    # See train_dreambooth_lora_sdxl.py for all of the training details.\n",
+    "    final_checkpoint_path, final_metrics = main(config[\"args\"])\n",
+    "\n",
+    "    # Upload final checkpoint to cloud. (Only the rank 0 worker will return a path here.)\n",
+    "    if final_checkpoint_path is not None:\n",
+    "        upload_to_cloud(local_path=final_checkpoint_path, cloud_uri=MODEL_CHECKPOINT_PATH)\n",
+    "\n",
+    "\n",
+    "trainer = TorchTrainer(\n",
+    "    train_fn_per_worker,\n",
+    "    # Pass command line arguments from the driver to the `config` dict of the `train_fn_per_worker`\n",
+    "    train_loop_config={\"args\": TRAINING_ARGS},\n",
+    "    scaling_config=ray.train.ScalingConfig(\n",
+    "        # Do data parallel training with A10G GPU workers\n",
+    "        num_workers=4, use_gpu=True, accelerator_type=\"A10G\"\n",
```

Comment:
This is a lot of workers for a tiny job, why not 1 or 2 at most?

Reply:
The original example uses an effective batch size of 4 by accumulating a batch size of 1 on a single device. Now, I just scale out the cluster to 4 gpus and don't do gradient accumulation to make the training run faster.

Finetuning requires around 100 steps to get decent results, which actually takes quite a bit of time with fewer workers. The current setup takes around 8 minutes.